### PR TITLE
support of supplementary alignments in phase

### PR DESCRIPTION
**Pull request by nkkarpov**
Adding support for supplementary alignments from issue https://github.com/whatshap/whatshap/issues/516 with flag --use-supplementary.

The supplementary alignment must have the same orientation as the primary and be within a distance of at most 100,000 from the primary alignment. This distance parameter can be adjusted using the --supplementary-distance option. The alignments may intersect. To form the new read from the alignments, the algorithm selects only positions where alignments fully agree with the allele of the variant and uses this allele.

### CODE REVIEW

File: whatshap/variants.py [hunk 1]
```diff
@@ -179,13 +202,60 @@ def _group_paired_reads(reads: Iterable[Read]) -> Iterator[List[Read]]:
...
+                                skip.add(variant.position)
+                        else:
+                            variants[variant.position] = variant
+                union_read = Read(primary.read.name,
+                                  primary.read.mapqs[0],
+                                  primary.read.source_id,
+                                  primary.read.sample_id,
+                                  reference_start,
+                                  primary.read.BX_tag, )
+                for k, v in variants.items():
```
Comment (marcelm) [inline whatshap/variants.py:243]: Instead of `k` and `v`, please use `position` and `variant`.

File: whatshap/variants.py [hunk 2]
```diff
@@ -179,13 +202,60 @@ def _group_paired_reads(reads: Iterable[Read]) -> Iterator[List[Read]]:
...
+                for read in group:
+                    if not read.is_supplementary:
+                        if primary is not None:
+                            raise ReadSetError(
+                                f"Read name {group[0].name!r} has more than two primary alignments.")
+                        primary = read
+                if primary is None:
+                    n_skipped += 1
+                    continue
+                reference_start = primary.reference_start
```
Comment (marcelm) [inline whatshap/variants.py:222]: Can the code from here maybe be factored out into a separate function (`merge_reads` or so)? There is a bit much going on and the indentation is a bit too deep.

File: whatshap/variants.py [hunk 3]
```diff
@@ -387,9 +469,9 @@ def split_cigar(cigar, i, consumed):
         else:
             left = cigar[:i]
         if consumed < middle_length:
-            right = [(middle_op, middle_length - consumed)] + cigar[i + 1 :]
```
Comment (marcelm) [inline whatshap/variants.py:390]: There are more formatting changes here and below that should not be made

**Review by marcelm — COMMENTED** [review]
Hi, as @tobiasmarschall said, this needs some tests. If you factor out a `merge_reads` function as I suggest, that would be a good candidate for a unit test.

I noticed some formatting changes that I guess your editor introduced. Please make sure to run `tox` locally to catch these issues. You can run `black` to fix the formatting (except between the `# fmt: off` and `# fmt: on` comments where you need to undo the changes some other way).

It would also be nice with some extra type annotations. We don’t run mypy automatically, yet, but it’s still nice for readability.

**Comment by nkkarpov** [comment]
Hello @marcelm, I've made the refactorings as you suggested and added unit tests for the new function. Please take a look when you have time. I found that the tests on genotyping failed on my local machine. However, these tests currently fail even if I clone from the main branch :) I didn't investigate further... I've also found that the test on haplotyping started to fail. From what I can see in the logs, some reads have two primary alignments in BAM files, which, as I understand, violates the BAM format. Is my understanding correct?

**Comment by marcelm** [comment]
Thanks!

> I found that the tests on genotyping failed on my local machine. However, these tests currently fail even if I clone from the main branch :)

Commit d3b6c77abd0179d1ac5d32bc54acc1e51a4eb17f, which is the most recent commit an the `main` branch, doesn’t fail for me locally nor on CI, see <https://github.com/whatshap/whatshap/actions/runs/8324657558>, so I assume it is something in this PR. But maybe it’s related to the haplotyping tests (see below).

> I've also found that the test on haplotyping started to fail. From what I can see in the logs, some reads have two primary alignments in BAM files, which, as I understand, violates the BAM format. Is my understanding correct?

I think the problem is that your code only allows one primary alignment for paired-end reads. For paired-end reads, each end can have a primary alignment.

This part of the SAM spec is relevant (and I agree and somewhat confusing):

> For each read/contig in a SAM file, it is required that one and only one line associated with the read satisfies ‘`FLAG & 0x900 == 0`’. This line is called the *primary line* of the read.

I think the point is that it says "each read", and that the SAM spec considers R1 and R2 to be two *different* reads although they have the same QNAME (they are considered to come from the same *template*, though). So there will be a primary alignment for R1 and one for R2.

(The SAM spec allows for reads with more than two "ends", but please ignore that part.)

**Comment by nkkarpov** [comment]
Thank you for the explanation! I fixed the issue and added more tests to cover this case.

File: whatshap/variants.py [hunk 4]
```diff
@@ -169,7 +188,67 @@ def _make_readset_from_grouped_reads(groups: Iterable[List[Read]]) -> ReadSet:
...
+        variants = dict()
+        skip = set()
+        for read in group:
+            if read.is_reverse != primary.is_reverse:
+                continue
+            if primary.distance(read) > distance_threshold:
+                continue
+            reference_start = min(reference_start, read.reference_start)
+            for variant in read.read:
+                if variant.position in variants.keys():
```
Comment (marcelm) [inline whatshap/variants.py:226]:
```suggestion
                if variant.position in variants:
```

**Comment by marcelm** [comment]
Nice, thanks. The unit tests are fine, but I wonder whether all corner cases are caught. Have you run this PR already on some real data? If it runs without crashing and maybe even gives the expected results, then I’d be more confident that the algorithm is ok.

Other than that, I think this only needs a changelog entry. I’ll wait with merging until tomorrow or so to give the other maintainers a chance to comment if they want.

**Comment by nkkarpov** [comment]
I tested the latest version on the internal data, and the results appear reasonable. This option reduces the number of phase blocks. In our data, we have only one primary alignment for each query name, and in this case, it doesn't affect us at all.

File: whatshap/cli/phase.py [hunk 5]
```diff
@@ -1140,6 +1146,12 @@ def add_arguments(parser):
     arg("--use-ped-samples", dest="use_ped_samples",
         action="store_true", default=False,
         help="Only work on samples mentioned in the provided PED file.")
+    arg("--use-supplementary", dest="use_supplementary", action="store_true", default=False,
+        help="Include supplementary alignments into reads")
```
Comment (marcelm) [inline whatshap/cli/phase.py:1150]:
```suggestion
        help="Use also supplementary alignments (default: ignore supplementary_ alignments)")
```

File: whatshap/cli/phase.py [hunk 6]
```diff
@@ -1140,6 +1146,12 @@ def add_arguments(parser):
...
+    arg("--supplementary-distance", dest="supplementary_distance_threshold", default=100_000,
+        help="Threshold for distance between primary and supplementary reads. (default: %(default)s)")
```
Comment (marcelm) [inline whatshap/cli/phase.py:1152]:
```suggestion
    arg("--supplementary-distance", metavar="DIST", dest="supplementary_distance_threshold", default=100_000,
        help="Skip supplementary alignments further than DIST bp away from the primary alignment (default: %(default)s)")
```

File: whatshap/variants.py [hunk 7]
```diff
@@ -180,13 +259,20 @@ def _group_paired_reads(reads: Iterable[Read]) -> Iterator[List[Read]]:
...
-            yield group
+            if len(group) > 1:
+                n_non_singleton += 1
+            read = ReadSetReader.create_read_from_group(group, distance_threshold)
+            if read is None:
+                n_skipped += 1
+            else:
+                yield [read]
+
+        logger.info(f"Number of non singleton groups: {n_non_singleton}")
```
Comment (marcelm) [inline whatshap/variants.py:274]:
```suggestion
        logger.info(f"Number of non-singleton groups: {n_non_singleton}")
```

**Review by marcelm — COMMENTED** [review]
I did a last round of review and made some language-related changes. I’ll just add them all and see whether I can undo the changes to `doc/develop.rst` as well, then this can be merged since none of the other maintainers objected.

**Comment by marcelm** [comment]
Nice, this has been on our wishlist for quite some time, thanks!

**Comment by marcelm** [comment]
Note I squashed the changes into a single commit to make it a bit cleaner.

**Comment by tobiasmarschall** [comment]
Hi @marcelm and @nkkarpov, I'm just returning from vacation today. Thanks for all your work on this!

Having a quick look at the code, my only comment is that I don't really love the `AlignedRead` class wrapped around `Read`. Is there a deeper reason for this, or was it just the reluctance to touch the `Read` class itself?

**Comment by marcelm** [comment]
I’m a bit undecided: Making `Read` itself have these extra attributes would allow us to get rid of the wrapper, but on the other hand, the attributes are only needed until supplementary and primary alignments have been merged, so it also feels ok to temporarily wrap them. That is, `AlignedRead` could be considered private to the `variants.py` module.

**Comment by tobiasmarschall** [comment]
Ok, given that there's pros/cons to both I'd vote to just merge the proposed version.

**Comment by nkkarpov** [comment]
You are right; the main reason I did this is because I didn't want to touch the C++/Cython code. One possible solution off the top of my head is to make ```AlignedRead``` a subclass of the ```Read``` class.

**Comment by leqi0001** [comment]
Hi,
Great feature! Is there a plan to allow supplemental alignment in polyphase function, or is it much more complicated to implement?
Thanks!

**Comment by marcelm** [comment]
Hi @leqi0001, I believe this should actually be quite easy to add since both `phase` and `polyphase` use the same underlying routines to read in the alignments. Can you please open a new issue? Comments on closed issues or PRs get lost easily.
